### PR TITLE
Print directive literals verbatim

### PR DESCRIPTION
**Pull request by josephfrazier**
This addresses https://github.com/prettier/prettier/issues/1555,
but doesn't seem to pass the AST_COMPARE=1 tests:

    AST_COMPARE=1 npm test -- tests/quotes -t strings

However, running `prettier --debug-check` on the relevant file *does*
work:

    prettier tests/quotes/strings.js --debug-check

### CODE REVIEW

**Comment by azz** [comment]
You'll need to update the snapshots with `npm t -- -u` to get the build passing.

File: src/printer.js [hunk 1]
```diff
@@ -3846,6 +3846,14 @@ function nodeStr(node, options) {
     ? alternate.quote
     : preferred.quote;
 
+  // Directives are exact code unit sequences, which means that you can't
+  // change the escape sequences they use.
+  // See https://github.com/prettier/prettier/issues/1555
+  // and https://tc39.github.io/ecma262/#directive-prologue
+  if (node.type === 'DirectiveLiteral') {
+    return raw;
+  }
```
Comment (vjeux) [inline src/printer.js]: Changing quotes if there are no escapes is totally fine. So we should only do this if shouldUseAlternateQuote is false or if it's true and numPreferredQuotes === 0 and numAlternateQuotes === 0

Comment (josephfrazier) [inline src/printer.js]: Good point, I just pushed up a fix at https://github.com/prettier/prettier/pull/1560/commits/5d338eb71baea71f0faddb14dd959bd0effa2458

It looks like that also fixes the tests, @azz (`npm test -- -u` didn't seem to do anything for me...)

File: src/printer.js [hunk 2]
```diff
@@ -3831,7 +3831,7 @@ function nodeStr(node, options) {
   const alternate = preferred === single ? double : single;
 
   let shouldUseAlternateQuote = false;
-  let canChangeDirectiveQuotes = true;
+  let canChangeDirectiveQuotes = !rawContent.includes('\\');
```
Comment (vjeux) [inline src/printer.js]: Why did you change this?

**Comment by vjeux** [comment]
Please add unit tests when you are changing behavior so that we don't regress in the future :) Thanks!

File: src/printer.js (hunk 2 above)
Comment (josephfrazier) [inline src/printer.js]: Before this change, the following was failing:

    prettier tests/quotes/strings.js --debug-check

Comment (vjeux) [inline src/printer.js]: Can you tell what was failing exactly?

Comment (josephfrazier) [inline src/printer.js]: Sure, it was the cases with escaped characters in the literals. Adding `--no-semi` makes it a little easier to see, so here's the full output of `prettier tests/quotes/strings.js --debug-check --no-semi`:

<details>

```
tests/quotes/strings.js
ast(input) !== ast(prettier(input))
Index: 
===================================================================
--- 	
+++ 	
@@ -73,5 +73,5 @@
         "value": {
           "type": "DirectiveLiteral",
-          "value": "\\'"
+          "value": "'"
         }
       },
@@ -80,5 +80,5 @@
         "value": {
           "type": "DirectiveLiteral",
-          "value": "\\'"
+          "value": "'"
         }
       },

Index: 
===================================================================
--- 	
+++ 	
@@ -9,15 +9,15 @@
 // Simple strings.
 "abc"
-'abc'
+"abc"
 
 // Escape.
-'\0'
+"\0"
 
 // Emoji.
-'🐶'
+"🐶"
 
 // Empty string.
 ""
-''
+""
 
 // Single double quote.
@@ -27,8 +27,8 @@
 // Single single quote.
 "'"
-'\''
+"'"
 
 // Unnecessary escapes.
-"\'"
+"'"
 '\"'
 
@@ -54,6 +54,6 @@
 
 // Single backslash.
-'\\'
 "\\"
+"\\"
 
 // Backslases.


```
</details>

**Comment by josephfrazier** [comment]
@vjeux, is it ok to ignore Flow/TypeScript in the tests for now, as done in https://github.com/prettier/prettier/pull/1560/commits/2aba6231271f6985a395c31e3df9323e8f3da115? I'm not sure what the policy is about merging changes that fix the default parser without also fixing the other ones (but I did see one instance where it's happened before: https://github.com/prettier/prettier/pull/602/commits/90bf93713c78a6a6b3f55e52d7be172ece9b56df#diff-0de18284f37da79ab8af4e4690919abaR1).

**Comment by vjeux** [comment]
This is a very edge case that is likely never going to be seen in practice, so might as well fix it everywhere or not.

**Comment by josephfrazier** [comment]
Good point, especially since that commit actually removes some tests, rather than just adding Babylon-specific ones.

**Comment by josephfrazier** [comment]
All that's left now is getting TypeScript working: https://travis-ci.org/prettier/prettier/jobs/230511014#L500

Unfortunately, it looks like the TS parser's AST doesn't indicate the "directive-ness" of a string literal, so it might need a bit more work. I'm going to keep looking into it.

**Comment by vjeux** [comment]
Still failing for TS it looks like.

image

**Comment by vjeux** [comment]
Feel free to remove the typescript test. I honestly don't think anyone is going to trip into this ever.

**Comment by josephfrazier** [comment]
Yeah... I briefly looked into getting `typescript-eslint-parser` to output directive information, but it seems like more trouble than it's worth at the moment, especially since TS support is a work-in-progress currently. Maybe I'll open an issue for it, at least to have it on the record.

File: src/printer.js [hunk 3]
```diff
@@ -181,6 +181,20 @@ function genericPrintNoParens(path, options, print, args) {
     case "EmptyStatement":
       return "";
     case "ExpressionStatement":
+      // Detect Flow-parsed directives
+      if (n.directive) {
+        return concat([
+          nodeStr(
+            {
```
Comment (vjeux) [inline src/printer.js]: Nooo, please don't fake objects that look like ast nodes. Instead it's better to add a boolean flag on the function argument saying if it's a directive or not and deal with it accordingly.

Comment (josephfrazier) [inline src/printer.js]: Oops, sorry about that! I thought this was good to go, but apparently not. I'll make sure to let you merge the next attempt :)

File: src/printer.js [hunk 4]
```diff
@@ -3831,6 +3845,7 @@ function nodeStr(node, options) {
   const alternate = preferred === single ? double : single;
 
   let shouldUseAlternateQuote = false;
+  let canChangeDirectiveQuotes = !rawContent.includes('\\');
```
Comment (vjeux) [inline src/printer.js]: I still don't understand why you need to check if there's a `\`. If you don't change quotes, then it shouldn't matter. If you do change quotes and there are no existing quotes, then having an escaped `\` is fine.

File: tests/directives/escaped.js [hunk 5]
```diff
@@ -0,0 +1,4 @@
+'\'';
+'\"';
+"\'";
```
Comment (vjeux) [inline tests/directives/escaped.js]: You don't need to escape this.

Comment (josephfrazier) [inline tests/directives/escaped.js]: I'm trying to test that even unnecessarily escaped characters are preserved as-is, since that's what "exact code unit sequence" means (as I understand it). From the [original bug report](https://github.com/prettier/prettier/issues/1555#issue-227206837):

> For example, prettier formats
> ```js
> " \" ";
> ```
> [as](https://prettier.github.io/prettier/#%7B%22content%22%3A%22%5C%22%20%5C%5C%5C%22%20%5C%22%3B%22%2C%22options%22%3A%7B%22printWidth%22%3A80%2C%22tabWidth%22%3A2%2C%22singleQuote%22%3Afalse%2C%22trailingComma%22%3A%22none%22%2C%22bracketSpacing%22%3Atrue%2C%22jsxBracketSameLine%22%3Afalse%2C%22parser%22%3A%22babylon%22%2C%22semi%22%3Atrue%2C%22useTabs%22%3Afalse%2C%22doc%22%3Afalse%7D%7D)
> ```
> ' " ';
> ```
> but these are not the same directive, because they do not contain the same exact code unit sequence.

Comment (vjeux) [inline tests/directives/escaped.js]: Oh I see, makes sense thanks!

File: src/printer.js (hunk 4 above)
Comment (josephfrazier) [inline src/printer.js]: You're right, we should ideally change to the preferred quote if possible. However, since changing the quote involves calling `makeString`, it could remove some of the unnecessary escapes, which would change the meaning of the directive (see https://github.com/prettier/prettier/pull/1560#discussion_r115609519). So, I just used the presence of any backslash at all as an indicator to leave the entire thing alone.

Hopefully, this is enough of an edge case that no one will actually run into it, as you mentioned in https://github.com/prettier/prettier/pull/1560#issuecomment-300296221. I'd still like to get it merged with the tests passing, though, so that we can set `AST_EXPLORE=1` on Travis to more rigorously test correctness.

File: src/printer.js (hunk 3 above)
Comment (josephfrazier) [inline src/printer.js]: Ok, I added https://github.com/josephfrazier/prettier/commit/c65cf2fe24d3c41cd92d4ee57fa8308317725e6a with what I think is the right idea here. Let me know what you think.

Note that I can't re-open the pull request, due to https://github.com/isaacs/github/issues/361. Once we get everything satisfactory, I can rebase and open a new PR with equivalent changes.

File: src/printer.js (hunk 4 above)
Comment (josephfrazier) [inline src/printer.js]: After a bit more thought, I decided to go ahead and make the logic a little more robust, so it covers the case you mentioned, about having an escaped `\` in the literal. See https://github.com/josephfrazier/prettier/commit/a52916f5755a8bc175d8fc519f318859cf3677ae

I also realized that the Github UI doesn't show any further changes on this branch past the ones that were merged, so I'll go ahead and open a new PR, linking back to this one.
